### PR TITLE
Throw exceptions instead of using status codes

### PR DESCRIPTION
**Pull request by dimaryaz**

### CODE REVIEW

File: plugins/nf-quilt/src/main/nextflow/quilt/QuiltProduct.groovy [hunk 1]
```diff
@@ -117,43 +117,29 @@ ${nextflow}
...
         if (session.isSuccess() || pkg.is_force()) {
-            try {
-                this.pubStatus = publish()
-            }
-            catch (Exception e) {
-                log.error("publish failed: ${e.getMessage()}", pkg.meta)
-            }
+            publish()
         } else {
             log.info("not publishing: ${pkg} [unsuccessful session]")
```
Comment (dimaryaz) [inline plugins/nf-quilt/src/main/nextflow/quilt/QuiltProduct.groovy:130]: I think this should throw an exception, too, but I don't know which one

File: plugins/nf-quilt/src/test/nextflow/quilt/QuiltProductTest.groovy [hunk 2]
```diff
@@ -141,15 +141,37 @@ class QuiltProductTest extends QuiltSpecification {
...
+        thrown(com.quiltdata.quiltcore.workflows.WorkflowException)
+
+        when:
+        makeWriteProduct(skip_meta) // no default metadata
+        then:
+        thrown(com.quiltdata.quiltcore.workflows.WorkflowException)
+
         // NOTE: push does NOT update local registry
+        expect:
         makeWriteProduct().pkg.install() // try to merge existing metadata
```
Comment (dimaryaz) [inline plugins/nf-quilt/src/test/nextflow/quilt/QuiltProductTest.groovy:169]: This actually fails!

But the existing test doesn't check the status code.

File: plugins/nf-quilt/src/main/nextflow/quilt/QuiltProduct.groovy (hunk 1 above)
Comment (drernie) [reply plugins/nf-quilt/src/main/nextflow/quilt/QuiltProduct.groovy:130]: Some options

[Sync­Failed­Exception](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/SyncFailedException.html) ✔

[Protocol­Exception](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/ProtocolException.html) ✔

[Server­Exception](https://docs.oracle.com/en/java/javase/11/docs/api/java.rmi/java/rmi/ServerException.html) ✔

Maybe define your own?

File: plugins/nf-quilt/src/test/nextflow/quilt/QuiltProductTest.groovy [hunk 3]
```diff
@@ -141,15 +141,37 @@ class QuiltProductTest extends QuiltSpecification {
...
-        makeWriteProduct(skip_meta).pubStatus == 1 // still fails on implicit metadata
+
+        when:
+        makeWriteProduct(skip_meta) // still fails on implicit metadata
+        then:
+        thrown(com.quiltdata.quiltcore.workflows.WorkflowException)
```
Comment (drernie) [inline plugins/nf-quilt/src/test/nextflow/quilt/QuiltProductTest.groovy:174]: I see you already defined one here.

quiltcore.package.PushException?

**Review by drernie — APPROVED** [review]

File: plugins/nf-quilt/src/main/nextflow/quilt/QuiltProduct.groovy (hunk 1 above)
Comment (dimaryaz) [reply plugins/nf-quilt/src/main/nextflow/quilt/QuiltProduct.groovy:130]: All of those are for very specific Java features, so doesn't seem right. They all inherit from `IOException`, so could just use that one.

But, `IOException` is for failures that the user has no control over (network failures, etc.) - while here, it looks like the user should've checked that the session is valid. If that's the case, then `IllegalStateException` might be more appropriate.

File: plugins/nf-quilt/src/test/nextflow/quilt/QuiltProductTest.groovy (hunk 3 above)
Comment (dimaryaz) [reply plugins/nf-quilt/src/test/nextflow/quilt/QuiltProductTest.groovy:174]: I actually never created our own `PushException`, but just used `java.io.IOException`. Still don't know what's best.

Comment (drernie) [reply plugins/nf-quilt/src/test/nextflow/quilt/QuiltProductTest.groovy:174]: java.io.IOException is fine. The user shouldn't really have control over the Session.

**Pull request merged**